pkg/store/storepb/inprocess.go

```diff
@@ -88,15 +90,13 @@ func (s *inProcessClientStream) CloseSend() error {
 
 func (s *inProcessClientStream) Recv() (*SeriesResponse, error) {
 	select {
-	case <-s.srv.ctx.Done():
-		return nil, s.srv.ctx.Err()
 	case r, ok := <-s.srv.recv:
 		if !ok {
 			return nil, io.EOF
 		}
 		return r, nil
-	case err := <-s.srv.err:
-		if err == nil {
+	case err, ok := <-s.srv.err:
```

Comment:
Could this lead to problems if `s.srv.err` is closed while we are continuously calling `Recv()` and there's something to read from `s.srv.recv`? In other words, from my understanding `select{}` chooses randomly if there are multiple possibilities so with `s.srv.err` closed we could prematurely return `io.EOF`, no? (https://go.dev/ref/spec#Select_statements, see second point). I think the same applies to the `case r, ok` branch. My suggestion would be to give a priority to `s.srv.recv`.

```go
select {
  case r, ok := <-s.srv.recv:
    if !ok {
      break
    }
    return r, nil
}

err, ok := <-s.srv.err
if ok {
  return nil, err
}
return nil, io.EOF
```

Reply:
I don't think that the race scenario you describe can happen here:

- The `s.srv.err` channel can only be closed when `s.srv.Series(in, inSrv)` returns. 
- The `Series()` method only returns after it has finished pushing data into the unbuffered `s.srv.recv` channel. 
- Consequently, `Series()` only returns once the last bits of data have been consumed by the `Recv()` caller.
- Making the suggested race impossible.

Regarding the proposed change to give priority to `s.srv.recv`, I see a potential issue if the `Series()` call returns an error: 

- `Recv()` will block on reading `s.srv.recv` which is not yet closed.
- `Series()` will block trying to write the error into the unbuffered `err` channel.
- This results in neither channel being closed, leading to a deadlock.

Do you agree @GiedriusS ?